<style  type="text/css"> 
</style>

<b><center>
<span style="font-size: 24pt; line-height: 1.2">
COMS W4111-002 (Fall 2021)<br>Introduction to Databases
</span>
</center></b>
</span><br>
<p>
<i><center>
<span style="font-size: 20pt; line-height: 1.2">
Homework 1: Programming - 10 Points
</span>
</center></i>

__Note:__ Please replace the information below with your last name, first name and UNI.<br><br>



<i>
<span style="font-size: 20pt; line-height: 1.2"; >
Yuming Tian, yt2777
</span>
</i>

## Introduction

### Objectives
This homework has you practice and build skill with:

- PART A: (1 point) Understanding relational databases
- PART B: (1 point) Understanding relational algebra
- PART C: (1 point) Cleaning data
- PART D: (1 point) Performing simple SQL queries to analyze the data.
- PART E: (6 points) CSVDataTable.py


__Note:__ The motivation for PART E may not be clear. The motivation will become clearer as the semester proceeds. The purpose of PART E is to get you started on programming in Python and manipulating data.



### Submission
1. File > Print Preview > Download as PDF
2. Upload .pdf and .ipynb to GradeScope
3. Upload CSVDataTable.py and CSVDataTable_Tests.py

**This assignment is due September 24, 11:59 pm ET**


### Collaboration
- You may use any information you get in TA or Prof. Ferguson's office hours, from lectures or from recitations.
- You may use information that you find on the web.
- You are NOT allowed to collaborate with other students outside of office hours.


# Part A: Written

1. What is a database management system?

    <i> It is a computerized data-keeping system that provides users with facilities to manage data as well as database structure </i>

2. What is a primary key and why is it important?

    <i> The primary key is an attribute in a table that can uniquely identify each tuple in that table. It is important because we can use primary key to declare relationships between tables. The primary key of one table could serve as a foreign key in another table and can be used to refer to the original data. </i>

3. Please explain the differences between SQL, MySQL Server and DataGrip?

    <i> SQL is a structured language that can be used to retreive/modify data in a database; MySQL, on the other hand, is the name of a specific database; DataGrip is a client that can be used to connect to both SQL and Non-SQL databases, which allows users to send queries and receive information from databases. </i>

4. What are 4 different types of DBMS table relationships, give a brief explanaition for each?

    <i> One to one: each tuple in table A has exactly one related tuple in table B, and so do tuples in table B </i><br>
    <i> One to many: each tuple in table A may have 0 to infinite related tuple(s) in table B, but each tuple in table B has at most one related tuple in table A </i><br>
    <i> Many to one: the reverse of one to many </i><br>
    <i> Many to many: each tuple in table A may have 0 to infinite related tuple(s) in table B, and so do tuples in table B </i>

5. What is an ER model? 

    <i> An ER model is a high-level graphical data model that defines data elements and their relationships for a specific system. </i>


6. Using Lucidchart draw an example of a logical ER model using Crow's Foot notation for Columbia classes. The entity types are:
    - Students, Professors, and Classes.
    - The relationships are:
        - A Class has exactly one Professor.
        - A Student has exactly one professor who is an _advisor._
        - A Professor may advise 0, 1 or many Students.
        - A Class has 0, 1 or many enrolled students.
        - A Student enrolls in 0, 1 or many Classes.
        
        
- You can define what you think are common attributes for each of the entity types. Do not define more than 5 or 6 attributes per entity type.


- In this example, explicitly show an example of a primary-key, foreign key, one-to-many relationship, and many-to-many relationship.


__Notes:__
- If you have not already done so, please register for a free account at Lucidchart.com. You can choose the option at the bottom of the left pane to add the ER diagram shapes.
- You can take a screen capture of you diagram and save in the zip directory that that contains you Jupyter notebook. Edit the following cell and replace "Boromir.jpg" with the name of the file containing your screenshot.
    
    

   <i> Use the following line to upload a photo of your Luicdchart. <i/>
   <img src="HW1_ER.png">

# Part B: Relational Algebra

You will use [the online relational calculator](https://dbis-uibk.github.io/relax/landing), choose the “Karlsruhe University of Applied Sciences” dataset.

An anti-join is a form of join with reverse logic. Instead of returning rows when there is a match (according to the join predicate) between the left and right side, an anti-join returns those rows from the left side of the predicate for which there is no match on the right. 

The Anti-Join Symbol is ▷.

Consider the following relational algebra expression and result.

/* (1) Set X = The set of classrooms in buildings Taylor or Watson. */

        X = σ building='Watson' ∨ building='Taylor' (classroom)

/* (2) Set Y =  The Anti-Join of department and X */

        Y = (department ▷ X)

/* (3) Display the rows in Y. */

        Y
        
   <img src="ra.png">


1. Find an alternate expression to (2) that computes the correct answer given X. Display the execution of your query below.

    <i> Query: Y = department - πdepartment.dept_name,department.building,department.budget (department ⨝ X) </i>
    <img src="RA_alt.png">


# Part C: Data Clean Up

## Please note: You MUST make a new schema using the lahmansdb_to_clean.sql file provided in the data folder. 

Use thelahmansdb_to_clean.sql file to make a new schema containing the raw data. The lahman database you created in Homework 0 has already been cleaned with all the constraints and will be used for Part D. Knowing how to clean data and add integrity constraints is very important which is why you go through the steps in part C. 

TLDR: If you use the HW0 lahman schema for this part you will get a lot of errors and recieve a lot of deductions. 

In [1]:
# You will need to follow instructions from HW 0 to make a new schema, import the data. 
# Connect to the unclean schema below by setting the database host, user ID and password.
%load_ext sql
%sql mysql+pymysql://root:MySQL_2020@192.168.1.245/lahmansdb_to_clean


Data cleanup: For each table we want you to clean, we have provided a list of changes you have to make. You can reference the cleaned lahman db for inspiration and guidance, but know that there are different ways to clean the data and you will be graded for your choice rationalization. You should make these changes through DataGrip's workbench's table editor and/or using SQL queries. In this part you will clean two tables: People and Batting. 


### You must have: 
- A brief explanation of why you think the change we requested is important.
- What change you made to the table.
- Any queries you used to make the changes, either the ones you wrote or the Alter statements provided by DataGrip's editor. 
- Executed the test statements we provided
- The cleaned table's new create statement (after you finish all the changes) 


### Overview of Changes:

People Table 

   0. Primary Key (Explanation is given, but you still must add the key to your table yourself)
   1. Empty strings to NULLs
   2. Column typing
   3. isDead column
   4. deathDate and birthDate column

Batting Table

   1. Empty strings to NULLs
   2. Column typing
   3. Primary Key
   4. Foreign Key
   

   
### How to make the changes: 

__Using the Table Editor:__

When you hit apply, a popup will open displaying the ALTER statments sql generates. Copy the sql provided first and paste it into this notebook. Then you can apply the changes. This means that you are NOT executing the ALTER statements through your notebook. 


1. Right click on the table > Modify Table...
<img src="modify.png" width="400" height="800">

2. Keys > press the + button > input the parameters > Execute 
    OR Keys > press the + button > input the parameters > copy and paste the script generated under "SQL Script" and paste into your notebook > Run the cell in jupyter notebook
<img src="pk.png" width="600" height="1200">




__Using sql queries:__

Copy paste any queries that you write manually into the notebook as well!


<hr style="height:2px">

## People Table

### 0) EXAMPLE: Add a Primary Key 

(Solutions are given but make sure you still do this step in workbench!)

#### Explanation

We want to add a Primary Key because we want to be able to uniquely identify rows within our data. A primary key is also an index, which allows us to locate data faster. 

#### Change

I added a Primary Key on the playerID column and made the datatype VARCHAR(15)

__Note:__ This is for demonstration purposes only. playerID __is not__ a primary key for fielding.

#### SQL

~~~~sql
ALTER TABLE `lahmansdb_to_clean`.`people`
CHANGE COLUMN `playerID` `playerID` VARCHAR(15) NOT NULL ,
ADD PRIMARY KEY (`playerID`);
~~~~

#### Tests

In [2]:
%sql SHOW KEYS FROM people WHERE Key_name = 'PRIMARY'
    

 * mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
1 rows affected.


[('people', 0, 'PRIMARY', 1, 'playerID', 'A', 20012, None, None, '', 'BTREE', '', '', 'YES', None)]

### 1) Convert all empty strings to NULL

#### Explanation

_It allows us to more effectively identifying empty values. For example, we can utilize operations such as isNull, natural join, etc_

#### Change

_Check all tuples by attributes. If there's an empty string, then update it's value to NULL_

#### SQL

~~~~sql
update `lahmansdb_to_clean`.`people` set birthYear = NULL where birthYear = '';
update `lahmansdb_to_clean`.`people` set birthMonth = NULL where birthMonth = '';
update `lahmansdb_to_clean`.`people` set birthDay = NULL where birthDay = '';
update `lahmansdb_to_clean`.`people` set birthCountry = NULL where birthCountry = '';
update `lahmansdb_to_clean`.`people` set birthState = NULL where birthState = '';
update `lahmansdb_to_clean`.`people` set birthCity = NULL where birthCity = '';
update `lahmansdb_to_clean`.`people` set deathYear = NULL where deathYear = '';
update `lahmansdb_to_clean`.`people` set deathMonth = NULL where deathMonth = '';
update `lahmansdb_to_clean`.`people` set deathDay = NULL where deathDay = '';
update `lahmansdb_to_clean`.`people` set deathCountry = NULL where deathCountry = '';
update `lahmansdb_to_clean`.`people` set deathState = NULL where deathState = '';
update `lahmansdb_to_clean`.`people` set deathCity = NULL where deathCity = '';
update `lahmansdb_to_clean`.`people` set nameFirst = NULL where nameFirst = '';
update `lahmansdb_to_clean`.`people` set nameLast = NULL where nameLast = '';
update `lahmansdb_to_clean`.`people` set nameGiven = NULL where nameGiven = '';
update `lahmansdb_to_clean`.`people` set weight = NULL where weight = '';
update `lahmansdb_to_clean`.`people` set height = NULL where height = '';
update `lahmansdb_to_clean`.`people` set bats = NULL where bats = '';
update `lahmansdb_to_clean`.`people` set throws = NULL where throws = '';
update `lahmansdb_to_clean`.`people` set debut = NULL where debut = '';
update `lahmansdb_to_clean`.`people` set finalGame = NULL where finalGame = '';
update `lahmansdb_to_clean`.`people` set retroID = NULL where retroID = '';
update `lahmansdb_to_clean`.`people` set bbrefID = NULL where bbrefID = '';
~~~~

#### Tests

In [3]:
%sql SELECT * FROM people WHERE birthState = ""


 * mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
0 rows affected.


[]

### 2) Change column datatypes to appropriate values (ENUM, INT, VARCHAR, DATETIME, ETC)

#### Explanation

_Matching the column datatypes with the actual data helps us to make comparisons. For example, using datetime for debut instead of text is more reasonable because then we can set a range and filter people whose debuts are within a certain period. Besides, using int for height instead of text allows us to compare which person is heigher; this can be useful in certain circumstances._

#### Change

_DATETIME: debut, finalGame<br>
VARCHAR(15): playerID, retroID, bbrefID<br>
TEXT: birthCountry, birthState, birthCity, deathCountry, deathState, deathCity, nameFirst, nameLast, nameGiven<br>
INT: birthYear, birthMonth, birthDay, deathYear, deathMonth, deathDay, weight, height<br>
ENUM: bats, throws_

#### SQL

~~~~sql
use lahmansdb_to_clean;
alter table people modify birthYear int null;
alter table people modify birthMonth int null;
alter table people modify birthDay int null;
alter table people modify deathYear int null;
alter table people modify deathMonth int null;
alter table people modify deathDay int null;
alter table people modify weight int null;
alter table people modify height int null;
alter table people modify bats enum('L', 'R', 'B', 'S') null;
alter table people modify throws enum('L', 'R', 'B', 'S') null;
alter table people modify debut datetime null;
alter table people modify finalGame datetime null;
alter table people modify retroID varchar(15) null;
alter table people modify bbrefID varchar(15) null;
~~~~

### 3) Add an isDead Column that is either 'Y' or 'N'
- Some things to think of: What data type should this column be? How do you know if the player is dead or not? Maybe you do not know if the player is dead.
- You do not need to make guesses about life spans, etc. Just apply a simple rule.

'Y' means the player is dead

'N' means the player is alive

#### Explanation

_Having this column can help us quickly determine wheter there're deathYear/deathMonth/etc. for the current tuple. If the record is 'N', then we don't even need to check those columns mentioned above. For players that we're not sure if they are alive or not, we could place a NUll in the isDead column. Otherwise, the value can be compute based on if the deathYear/etc. exist._

#### Change

_Ceate a new column called isDead with type enum('Y', 'N'). Update each tuple to check if deathYear is NULL. If it is NULL, then place a 'Y'(true) in the isDead column, 'N'(false) otherwise._

#### SQL

~~~~sql
use lahmansdb_to_clean;
alter table people add isDead enum('Y', 'N') null;
update people set isDead = 'Y';
update people set isDead = 'N' where deathYear is null;
~~~~

#### Tests

In [4]:
%sql SELECT * FROM people WHERE isDead = "N" limit 10

 * mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
10 rows affected.


[('aardsda01', 1981, 12, 27, 'USA', 'CO', 'Denver', None, None, None, None, None, None, 'David', 'Aardsma', 'David Allan', 215, 75, 'R', 'R', datetime.datetime(2004, 4, 6, 0, 0), datetime.datetime(2015, 8, 23, 0, 0), 'aardd001', 'aardsda01', 'N', datetime.datetime(1981, 12, 27, 0, 0), None),
 ('aaronha01', 1934, 2, 5, 'USA', 'AL', 'Mobile', None, None, None, None, None, None, 'Hank', 'Aaron', 'Henry Louis', 180, 72, 'R', 'R', datetime.datetime(1954, 4, 13, 0, 0), datetime.datetime(1976, 10, 3, 0, 0), 'aaroh101', 'aaronha01', 'N', datetime.datetime(1934, 2, 5, 0, 0), None),
 ('aasedo01', 1954, 9, 8, 'USA', 'CA', 'Orange', None, None, None, None, None, None, 'Don', 'Aase', 'Donald William', 190, 75, 'R', 'R', datetime.datetime(1977, 7, 26, 0, 0), datetime.datetime(1990, 10, 3, 0, 0), 'aased001', 'aasedo01', 'N', datetime.datetime(1954, 9, 8, 0, 0), None),
 ('abadan01', 1972, 8, 25, 'USA', 'FL', 'Palm Beach', None, None, None, None, None, None, 'Andy', 'Abad', 'Fausto Andres', 184, 73, 'L', 'L', datetime.datetime(2001, 9, 10, 0, 0), datetime.datetime(2006, 4, 13, 0, 0), 'abada001', 'abadan01', 'N', datetime.datetime(1972, 8, 25, 0, 0), None),
 ('abadfe01', 1985, 12, 17, 'D.R.', 'La Romana', 'La Romana', None, None, None, None, None, None, 'Fernando', 'Abad', 'Fernando Antonio', 235, 74, 'L', 'L', datetime.datetime(2010, 7, 28, 0, 0), datetime.datetime(2019, 9, 28, 0, 0), 'abadf001', 'abadfe01', 'N', datetime.datetime(1985, 12, 17, 0, 0), None),
 ('abbotgl01', 1951, 2, 16, 'USA', 'AR', 'Little Rock', None, None, None, None, None, None, 'Glenn', 'Abbott', 'William Glenn', 200, 78, 'R', 'R', datetime.datetime(1973, 7, 29, 0, 0), datetime.datetime(1984, 8, 8, 0, 0), 'abbog001', 'abbotgl01', 'N', datetime.datetime(1951, 2, 16, 0, 0), None),
 ('abbotje01', 1972, 8, 17, 'USA', 'GA', 'Atlanta', None, None, None, None, None, None, 'Jeff', 'Abbott', 'Jeffrey William', 190, 74, 'R', 'L', datetime.datetime(1997, 6, 10, 0, 0), datetime.datetime(2001, 9, 29, 0, 0), 'abboj002', 'abbotje01', 'N', datetime.datetime(1972, 8, 17, 0, 0), None),
 ('abbotji01', 1967, 9, 19, 'USA', 'MI', 'Flint', None, None, None, None, None, None, 'Jim', 'Abbott', 'James Anthony', 200, 75, 'L', 'L', datetime.datetime(1989, 4, 8, 0, 0), datetime.datetime(1999, 7, 21, 0, 0), 'abboj001', 'abbotji01', 'N', datetime.datetime(1967, 9, 19, 0, 0), None),
 ('abbotku01', 1969, 6, 2, 'USA', 'OH', 'Zanesville', None, None, None, None, None, None, 'Kurt', 'Abbott', 'Kurt Thomas', 180, 71, 'R', 'R', datetime.datetime(1993, 9, 7, 0, 0), datetime.datetime(2001, 4, 13, 0, 0), 'abbok002', 'abbotku01', 'N', datetime.datetime(1969, 6, 2, 0, 0), None),
 ('abbotky01', 1968, 2, 18, 'USA', 'MA', 'Newburyport', None, None, None, None, None, None, 'Kyle', 'Abbott', 'Lawrence Kyle', 200, 76, 'L', 'L', datetime.datetime(1991, 9, 10, 0, 0), datetime.datetime(1996, 8, 24, 0, 0), 'abbok001', 'abbotky01', 'N', datetime.datetime(1968, 2, 18, 0, 0), None)]

### 4) Add a deathDate and birthDate column
Some things to think of: What do you do if you are missing information? What datatype should this column be?

You have to create this column from other columns in the table.

#### Explanation

_Having deathDate and birthDate columns in the DATETIME type allows us to easily compare values (ex. who was born earlier) without collecting info from three columns and using complex if statements._

#### Change

_For birthDate, we can compute it with values from birthYear, birthMonth, and birthDay; similar for deathDate. If any of the year/month/day value is missing, then it means that we don't have sufficient information, and therefore the date should be set to NULL as well._

#### SQL

~~~~sql
use lahmansdb_to_clean;
alter table people add birthDate datetime null;
alter table people add deathDate datetime null;
update people set birthDate = null, deathDate = null;
update people set birthDate = concat(birthYear, '-' , birthMonth, '-' , birthDay) where birthYear is not null;
update people set deathDate = concat(deathYear, '-' , deathMonth, '-' , deathDay) where deathYear is not null;
~~~~

#### Tests

In [5]:
%sql SELECT deathDate FROM people WHERE deathDate >= '2005-01-01' ORDER BY deathDate ASC LIMIT 10;

 * mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
10 rows affected.


[(datetime.datetime(2005, 1, 4, 0, 0),),
 (datetime.datetime(2005, 1, 7, 0, 0),),
 (datetime.datetime(2005, 1, 9, 0, 0),),
 (datetime.datetime(2005, 1, 10, 0, 0),),
 (datetime.datetime(2005, 1, 21, 0, 0),),
 (datetime.datetime(2005, 1, 22, 0, 0),),
 (datetime.datetime(2005, 1, 31, 0, 0),),
 (datetime.datetime(2005, 2, 4, 0, 0),),
 (datetime.datetime(2005, 2, 8, 0, 0),),
 (datetime.datetime(2005, 2, 11, 0, 0),)]

In [6]:
%sql SELECT birthDate FROM people WHERE birthDate <= '1965-01-01' ORDER BY birthDate ASC LIMIT 10;


 * mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
10 rows affected.


[(datetime.datetime(1820, 4, 17, 0, 0),),
 (datetime.datetime(1824, 10, 5, 0, 0),),
 (datetime.datetime(1832, 9, 17, 0, 0),),
 (datetime.datetime(1832, 10, 23, 0, 0),),
 (datetime.datetime(1835, 1, 10, 0, 0),),
 (datetime.datetime(1836, 2, 29, 0, 0),),
 (datetime.datetime(1837, 12, 26, 0, 0),),
 (datetime.datetime(1838, 3, 10, 0, 0),),
 (datetime.datetime(1838, 7, 16, 0, 0),),
 (datetime.datetime(1838, 8, 27, 0, 0),)]

### Final CREATE Statement

To find the create statement:

- Right click on the table name in workbench
- Select 'Copy to Clipboard'
- Select 'Create Statement' 

The create statement will now be copied into your clipboard and can be pasted into the cell below. 

~~~~sql
create table lahmansdb_to_clean.people
(
	playerID varchar(15) not null
		primary key,
	birthYear int null,
	birthMonth int null,
	birthDay int null,
	birthCountry varchar(20) null,
	birthState text null,
	birthCity text null,
	deathYear int null,
	deathMonth int null,
	deathDay int null,
	deathCountry text null,
	deathState text null,
	deathCity text null,
	nameFirst text null,
	nameLast text null,
	nameGiven text null,
	weight int null,
	height int null,
	bats enum('L', 'R', 'B', 'S') null,
	throws enum('L', 'R', 'B', 'S') null,
	debut datetime null,
	finalGame datetime null,
	retroID varchar(15) null,
	bbrefID varchar(15) null,
	isDead enum('Y', 'N') null,
	birthDate datetime null,
	deathDate datetime null
);
~~~~

<hr style="height:2px">

## Batting Table

### 1) Convert all empty strings to NULL

#### Explanation

_Similar to the People table, converting empty strings to NULL allows us to more effectively identifying empty values. For example, we can utilize operations such as isNull, natural join, etc_

#### Change

_Check all tuples and replace any empty string found._

#### SQL

~~~~sql
use lahmansdb_to_clean;
update batting set playerID = null where playerID = '';
update batting set yearID = null where yearID = '';
update batting set stint = null where stint = '';
update batting set teamID = null where teamID = '';
update batting set lgID = null where lgID = '';
update batting set G = null where G = '';
update batting set AB = null where AB = '';
update batting set R = null where R = '';
update batting set H = null where H = '';
update batting set `2B` = null where `2B` = '';
update batting set `3B` = null where `3B` = '';
update batting set HR = null where HR = '';
update batting set RBI = null where RBI = '';
update batting set SB = null where SB = '';
update batting set CS = null where CS = '';
update batting set BB = null where BB = '';
update batting set SO = null where SO = '';
update batting set IBB = null where IBB = '';
update batting set HBP = null where HBP = '';
update batting set SH = null where SH = '';
update batting set SF = null where SF = '';
update batting set GIDP = null where GIDP = '';
~~~~

#### Tests

In [7]:
%sql SELECT count(*) FROM lahmansdb_to_clean.batting where RBI is NULL;

 * mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
1 rows affected.


[(756,)]

### 2) Change column datatypes to appropriate values (ENUM, INT, VARCHAR, DATETIME, ETC)

#### Explanation

_Appropriate data types help us to more easily make comparisons or filter data. Similar to the People table._

#### Change

_Text (no changes): teamID, lgID<br>
varchar(15): playerID - since it is a string with fixed length<br>
int - includes year and score-related columns: yearID, stint, G, AB, R, H, 2B, 3B, HR, RBI, SB, CS, BB, SO, IBB, HBP, SH, SF, GIDP_

#### SQL

~~~~sql
use lahmansdb_to_clean;
alter table batting modify playerID varchar(15) not null;
alter table batting modify yearID int null;
alter table batting modify stint int null;
alter table batting modify G int null;
alter table batting modify AB int null;
alter table batting modify R int null;
alter table batting modify H int null;
alter table batting modify `2B` int null;
alter table batting modify `3B` int null;
alter table batting modify HR int null;
alter table batting modify RBI int null;
alter table batting modify SB int null;
alter table batting modify CS int null;
alter table batting modify BB int null;
alter table batting modify SO int null;
alter table batting modify IBB int null;
alter table batting modify HBP int null;
alter table batting modify SH int null;
alter table batting modify SF int null;
alter table batting modify GIDP int null;
~~~~

### 3) Add a Primary Key
Two options for the Primary Key:

- Composite Key: playerID, yearID, stint
- Covering Key (Index): playerID, yearID, stint, teamID


#### Choice

_playerID, yearID, stint_

#### Explanation


_These three values are sufficient to uniquely identify each tuple in this table, given that each player cannot join different matches at the same time._

#### SQL

~~~~sql
use lahmansdb_to_clean;
alter table batting add primary key (playerID, yearID, stint);
~~~~

#### Test

In [8]:
%sql SHOW KEYS FROM batting WHERE Key_name = 'PRIMARY' and Column_name = 'playerID'

 * mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
1 rows affected.


[('batting', 0, 'PRIMARY', 1, 'playerID', 'A', 19840, None, None, '', 'BTREE', '', '', 'YES', None)]

### 4) Add a foreign key on playerID between the People and Batting Tables
Note: Two people in the batting table do not exist in the people table. How should you handle this issue?

#### Explanation

_By setting playerID as a foreign key refering to the People table, we can ensure that each player shown in the Batting table is acually existed in the People table._

#### Change

_set playerID as a foreign key_

#### SQL

~~~~sql
use lahmansdb_to_clean;
alter table batting add foreign key (playerID) references people(playerID);
~~~~

#### Tests

In [9]:
%sql Select playerID from batting where playerID not in (select playerID from people);

 * mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
0 rows affected.


[]

### Final CREATE Statement

To find the create statement:

- Right click on the table name in workbench
- Select 'Copy to Clipboard'
- Select 'Create Statement' 

The create statement will now be copied into your clipboard and can be pasted into the cell below. 

~~~~sql
create table lahmansdb_to_clean.batting
(
	playerID varchar(15) not null,
	yearID int not null,
	stint int not null,
	teamID text null,
	lgID text null,
	G int null,
	AB int null,
	R int null,
	H int null,
	`2B` int null,
	`3B` int null,
	HR int null,
	RBI int null,
	SB int null,
	CS int null,
	BB int null,
	SO int null,
	IBB int null,
	HBP int null,
	SH int null,
	SF int null,
	GIDP int null,
	primary key (playerID, yearID, stint),
	constraint batting_ibfk_1
		foreign key (playerID) references lahmansdb_to_clean.people (playerID)
);


~~~~

# Part D: SQL Queries

NOTE: You must use the CLEAN lahman schema provided in HW0 for the queries below to ensure your answers are consistent with the solutions. 

In [10]:
%reload_ext sql
%sql mysql+pymysql://root:MySQL_2020@192.168.1.245/lahmansbaseballdb

## Question 0
What is the average salary in baseball history?

In [11]:
%sql SELECT avg(salary) from salaries;

 * mysql+pymysql://root:***@192.168.1.245/lahmansbaseballdb
   mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
1 rows affected.


[(2085634.053125473,)]

## Question 1
Select the players with a first name of Sam who were born in the United States and attended college.

Include their first name, last name, playerID, school ID, yearID and birth state. Limit 10

Hint: Use a Join between People and CollegePlaying

In [12]:
%sql select nameFirst, nameLast, playerID, schoolID, yearID, birthState from people natural join collegeplaying where nameFirst = 'Sam' and birthCountry = 'USA' limit 10;

 * mysql+pymysql://root:***@192.168.1.245/lahmansbaseballdb
   mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
10 rows affected.


[('Sam', 'Barnes', 'barnesa01', 'auburn', 1918, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1919, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1920, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1921, 'AL'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1956, 'NC'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1957, 'NC'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1958, 'NC'),
 ('Sam', 'Bowen', 'bowensa02', 'gacoast', 1971, 'GA'),
 ('Sam', 'Bowen', 'bowensa02', 'gacoast', 1972, 'GA'),
 ('Sam', 'Brown', 'brownsa01', 'grovecity', 1899, 'PA')]

## Question 2
Update all entries with full_name Columbia University to 'Columbia University in the City of New York' in the Schools table. Then select the row.


In [13]:
%sql update schools set name_full = 'Columbia University in the City of New York' where schoolID = 'columbia';
%sql select * from schools where schoolID = 'columbia';

 * mysql+pymysql://root:***@192.168.1.245/lahmansbaseballdb
   mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
1 rows affected.
 * mysql+pymysql://root:***@192.168.1.245/lahmansbaseballdb
   mysql+pymysql://root:***@192.168.1.245/lahmansdb_to_clean
1 rows affected.


[('columbia', 'Columbia University in the City of New York', 'New York', 'NY', 'USA')]

# Part E: CSVDataTable


## i. Conceptual Questions
The purpose of this homework is to teach you the behaviour of SQL Databases by asking you to implement functions that will model the behaviour of a real database with CSVDataTable. You will mimic a SQL Database using CSV files. 

Read through the scaffolding code provided in the CSVDataTable folder first to understand and answer the following conceptual questions.

1. Given this SQL statement:
    
        SELECT nameFirst, nameLast FROM people WHERE playerID = collied01
   
   If you run find_by_primary_key() on this statement, what are key_fields and field_list? 

    <i> key_fields should be ['collied01']; field_list should be ['nameFirst', 'nameLast'] </i>

2. What should be checked when you are trying to INSERT a new row into a table with a PK?

    <i> We need to make sure the PK of the row is not NULL and hasn't been used in the table (in other words, must be unique). </i>

3. What should be checked when you are trying to UPDATE a row in a table with a PK?

    <i> We should first try to find the row(s) that we want to update, since it is possible that they don't exist in the table. If we cannot find any row matching the conditions, no action should be performed. Otherwise, we modify the columns after making sure the new values don't violate the PK constraints (similar to the previous question, unique and not NULL). </i>

## ii. Coding


You are responsible for implementing and testing two classes in Python: CSVDataTable, BaseDataTable.
The python files and data can be found in the assignment under Courseworks. 

We have already given you **find_by_template(self, template, field_list=None, limit=None, offset=None, order_by=None)**
Use this as a jumping off point for the rest of your functions.

Methods to complete:

CSVDataTable.py
- find_by_primary_key(self, key_fields, field_list=None)
- delete_by_key(self, key_fields)
- delete_by_template(self, template)
- update_by_key(self, key_fields, new_values)
- update_by_template(self, template, new_values)
- insert(self, new_record)
CSV_table_tests.py
- You must test all methods. You will have to write these tests yourself. 
- You must test your methods on the People and Batting table.

If you do not include tests and tests outputs 50% of this section's points will be deducted at the start




## iii. Testing

Please copy the text from the output of your tests and paste it below:


DEBUG:root:CSVDataTable.__init__: data = {
  "table_name": "People",
  "connect_info": {
    "directory": "/Users/yumingtian/Documents/Columbia/W4111/HW1/HW1/Data/Baseball",
    "file_name": "People.csv"
  },
  "key_columns": [
    "playerID"
  ],
  "debug": true
}
DEBUG:root:CSVDataTable._load: Loaded 19619 rows
############################################################
#                      Table People                        #
############################################################

find_by_primary_key(): Known Record
[{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}]

find_by_primary_key(): Unknown Record
[]

find_by_template(): Known Template
[{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}]

delete_by_key(): Known Record
1

delete_by_key(): Unknown Record
0

delete_by_template: Known Template
1

delete_by_template: Unknown Template
0

update_by_key(): Known Record
1

update_by_key(): Unknown Record
0

update_by_key(): Bad PK
0

update_by_template(): Known Record
1

update_by_template(): Unknown Record
0

update_by_template(): Bad PK
0

insert(): Valid Record
[{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}]

insert(): Invalid Record
[]

insert(): Bad PK
[]

DEBUG:root:CSVDataTable.__init__: data = {
  "table_name": "Batting",
  "connect_info": {
    "directory": "/Users/yumingtian/Documents/Columbia/W4111/HW1/HW1/Data/Baseball",
    "file_name": "Batting.csv"
  },
  "key_columns": [
    "playerID",
    "yearID",
    "stint"
  ],
  "debug": true
}
DEBUG:root:CSVDataTable._load: Loaded 105861 rows
############################################################
#                      Table Batting                       #
############################################################

find_by_primary_key(): Known Record
[{'playerID': 'abercda01', 'yearID': '1871', 'stint': '1', 'teamID': 'TRO', 'lgID': 'NA', 'G': '1', 'AB': '4', 'R': '0', 'H': '0', '2B': '0', '3B': '0', 'HR': '0', 'RBI': '0', 'SB': '0', 'CS': '0', 'BB': '0', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}]

find_by_primary_key(): Unknown Record
[]

find_by_template(): Known Template
[{'playerID': 'abercda01', 'yearID': '1871', 'stint': '1', 'teamID': 'TRO', 'lgID': 'NA', 'G': '1', 'AB': '4', 'R': '0', 'H': '0', '2B': '0', '3B': '0', 'HR': '0', 'RBI': '0', 'SB': '0', 'CS': '0', 'BB': '0', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}]

delete_by_key(): Known Record
1

delete_by_key(): Unknown Record
0

delete_by_template: Known Template
1

delete_by_template: Unknown Template
0

update_by_key(): Known Record
1

update_by_key(): Unknown Record
0

update_by_key(): Bad PK
0

update_by_template(): Known Record
1

update_by_template(): Unknown Record
0

update_by_template(): Bad PK
0

insert(): Valid Record
[{'playerID': 'abercda01', 'yearID': '1871', 'stint': '1', 'teamID': 'TRO', 'lgID': 'NA', 'G': '1', 'AB': '4', 'R': '0', 'H': '0', '2B': '0', '3B': '0', 'HR': '0', 'RBI': '0', 'SB': '0', 'CS': '0', 'BB': '0', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}]

insert(): Invalid Record
[]

insert(): Bad PK
[]